java.lang.IllegalArgumentException: Cannot add two delimiter processors for char '~' and minimum length 1

Using version 0.20.0 of commonmark-java with strikethrough extension leads to this exception at runtime:

> Caused by: java.lang.IllegalArgumentException: Cannot add two delimiter processors for char '~' and minimum length 1
> 	at org.commonmark.internal.StaggeredDelimiterProcessor.add(StaggeredDelimiterProcessor.java:54) ~[commonmark-0.20.0.jar:na]
> 	at org.commonmark.internal.InlineParserImpl.addDelimiterProcessors(InlineParserImpl.java:92) ~[commonmark-0.20.0.jar:na]
> 	at org.commonmark.internal.InlineParserImpl.calculateDelimiterProcessors(InlineParserImpl.java:69) ~[commonmark-0.20.0.jar:na]
> 	at org.commonmark.internal.InlineParserImpl.<init>(InlineParserImpl.java:39) ~[commonmark-0.20.0.jar:na]
> 	at org.commonmark.parser.Parser$Builder$1.create(Parser.java:272) ~[commonmark-0.20.0.jar:na]
> 	at org.commonmark.parser.Parser.<init>(Parser.java:45) ~[commonmark-0.20.0.jar:na]
> 	at org.commonmark.parser.Parser.<init>(Parser.java:28) ~[commonmark-0.20.0.jar:na]
> 	at org.commonmark.parser.Parser$Builder.build(Parser.java:128) ~[commonmark-0.20.0.jar:na]
> 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:na]
> 	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77) ~[na:na]
> 	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:na]
> 	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499) ~[na:na]
> 	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480) ~[na:na]
> 	at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:211) ~[spring-beans-5.3.23.jar:5.3.23]
> 	... 47 common frames omitted

Gradle dependencies used are:

> implementation 'org.commonmark:commonmark:0.20.0'
> implementation 'org.commonmark:commonmark-ext-autolink:0.20.0'
> implementation 'org.commonmark:commonmark-ext-gfm-tables:0.20.0'
> implementation 'org.commonmark:commonmark-ext-gfm-strikethrough:0.20.0'
> implementation 'org.commonmark:commonmark-ext-heading-anchor:0.20.0'
> implementation 'org.commonmark:commonmark-ext-image-attributes:0.20.0'

The renderer is initialized as follows:

```java
final List<Extension> extensions =
    Arrays.asList(
        AutolinkExtension.create(),
        HeadingAnchorExtension.create(),
        ImageAttributesExtension.create(),
        UnderlineExtension.create(),
        StrikethroughExtension.create(),
        SubscriptExtension.create(),
        SupscriptExtension.create(),
        TablesExtension.create());

this.markdownParser =
    Parser.builder().extensions(extensions).build();
this.markdownHtmlRenderer =
    HtmlRenderer.builder()
        .escapeHtml(true)
        .extensions(extensions)
        .nodeRendererFactory(VideoEmbedNodeRenderer::new)
        .sanitizeUrls(true)
        .build();
this.markdownPlaintextRenderer = TextContentRenderer.builder().extensions(extensions).stripNewlines(true).build();
```

Removing `StrikethroughExtension.create()` from the list of extensions or downgrading to version 0.19.0 fixes the issue.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,14 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html),
 with the exception that 0.x versions can break between minor versions.
 
+## [Unreleased]
+### Added
+- GitHub strikethrough: With the previous version we adjusted the
+  extension to also accept the single tilde syntax. But if you use
+  another extension that uses the single tilde syntax, you will get a
+  conflict. To avoid that, `StrikethroughExtension` can now be
+  configured to require two tildes like before, see Javadoc.
+
 ## [0.20.0] - 2022-10-20
 ### Fixed
 - GitHub tables: A single pipe (optional whitespace) now ends a table
@@ -371,7 +379,7 @@ API breaking changes (caused by changes in spec):
 Initial release of commonmark-java, a port of commonmark.js with extensions
 for autolinking URLs, GitHub flavored strikethrough and tables.
 
-
+[Unreleased]: https://github.com/commonmark/commonmark-java/compare/commonmark-parent-0.20.0...HEAD
 [0.20.0]: https://github.com/commonmark/commonmark-java/compare/commonmark-parent-0.19.0...commonmark-parent-0.20.0
 [0.19.0]: https://github.com/commonmark/commonmark-java/compare/commonmark-parent-0.18.2...commonmark-parent-0.19.0
 [0.18.2]: https://github.com/commonmark/commonmark-java/compare/commonmark-parent-0.18.1...commonmark-parent-0.18.2

--- a/commonmark-ext-gfm-strikethrough/src/main/java/org/commonmark/ext/gfm/strikethrough/StrikethroughExtension.java
+++ b/commonmark-ext-gfm-strikethrough/src/main/java/org/commonmark/ext/gfm/strikethrough/StrikethroughExtension.java
@@ -30,11 +30,13 @@ import org.commonmark.renderer.NodeRenderer;
  * <p>
  * If you have another extension that only uses a single tilde ({@code ~}) syntax, you will have to configure this
  * {@link StrikethroughExtension} to only accept the double tilde syntax, like this:
+ * </p>
  * <pre>
  *     {@code
  *     StrikethroughExtension.builder().requireTwoTildes(true).build();
  *     }
  * </pre>
+ * <p>
  * If you don't do that, there's a conflict between the two extensions and you will get an
  * {@link IllegalArgumentException} when constructing the parser.
  * </p>

--- a/commonmark-ext-gfm-strikethrough/src/main/java/org/commonmark/ext/gfm/strikethrough/StrikethroughExtension.java
+++ b/commonmark-ext-gfm-strikethrough/src/main/java/org/commonmark/ext/gfm/strikethrough/StrikethroughExtension.java
@@ -14,29 +14,57 @@ import org.commonmark.parser.Parser;
 import org.commonmark.renderer.NodeRenderer;
 
 /**
- * Extension for GFM strikethrough using ~~ (GitHub Flavored Markdown).
+ * Extension for GFM strikethrough using {@code ~} or {@code ~~} (GitHub Flavored Markdown).
+ * <p>Example input:</p>
+ * <pre>{@code ~foo~ or ~~bar~~}</pre>
+ * <p>Example output (HTML):</p>
+ * <pre>{@code <del>foo</del> or <del>bar</del>}</pre>
  * <p>
- * Create it with {@link #create()} and then configure it on the builders
+ * Create the extension with {@link #create()} and then add it to the parser and renderer builders
  * ({@link org.commonmark.parser.Parser.Builder#extensions(Iterable)},
  * {@link HtmlRenderer.Builder#extensions(Iterable)}).
  * </p>
  * <p>
  * The parsed strikethrough text regions are turned into {@link Strikethrough} nodes.
  * </p>
+ * <p>
+ * If you have another extension that only uses a single tilde ({@code ~}) syntax, you will have to configure this
+ * {@link StrikethroughExtension} to only accept the double tilde syntax, like this:
+ * <pre>
+ *     {@code
+ *     StrikethroughExtension.builder().requireTwoTildes(true).build();
+ *     }
+ * </pre>
+ * If you don't do that, there's a conflict between the two extensions and you will get an
+ * {@link IllegalArgumentException} when constructing the parser.
+ * </p>
  */
 public class StrikethroughExtension implements Parser.ParserExtension, HtmlRenderer.HtmlRendererExtension,
         TextContentRenderer.TextContentRendererExtension {
 
-    private StrikethroughExtension() {
+    private final boolean requireTwoTildes;
+
+    private StrikethroughExtension(Builder builder) {
+        this.requireTwoTildes = builder.requireTwoTildes;
     }
 
+    /**
+     * @return the extension with default options
+     */
     public static Extension create() {
-        return new StrikethroughExtension();
+        return builder().build();
+    }
+
+    /**
+     * @return a builder to configure the behavior of the extension
+     */
+    public static Builder builder() {
+        return new Builder();
     }
 
     @Override
     public void extend(Parser.Builder parserBuilder) {
-        parserBuilder.customDelimiterProcessor(new StrikethroughDelimiterProcessor());
+        parserBuilder.customDelimiterProcessor(new StrikethroughDelimiterProcessor(requireTwoTildes));
     }
 
     @Override
@@ -57,5 +85,27 @@ public class StrikethroughExtension implements Parser.ParserExtension, HtmlRende
                 return new StrikethroughTextContentNodeRenderer(context);
             }
         });
+    }
+
+    public static class Builder {
+
+        private boolean requireTwoTildes = false;
+
+        /**
+         * @param requireTwoTildes Whether two tilde characters ({@code ~~}) are required for strikethrough or whether
+         * one is also enough. Default is {@code false}; both a single tilde and two tildes can be used for strikethrough.
+         * @return {@code this}
+         */
+        public Builder requireTwoTildes(boolean requireTwoTildes) {
+            this.requireTwoTildes = requireTwoTildes;
+            return this;
+        }
+
+        /**
+         * @return a configured extension
+         */
+        public Extension build() {
+            return new StrikethroughExtension(this);
+        }
     }
 }

--- a/commonmark-ext-gfm-strikethrough/src/main/java/org/commonmark/ext/gfm/strikethrough/internal/StrikethroughDelimiterProcessor.java
+++ b/commonmark-ext-gfm-strikethrough/src/main/java/org/commonmark/ext/gfm/strikethrough/internal/StrikethroughDelimiterProcessor.java
@@ -10,6 +10,16 @@ import org.commonmark.parser.delimiter.DelimiterRun;
 
 public class StrikethroughDelimiterProcessor implements DelimiterProcessor {
 
+    private final boolean requireTwoTildes;
+
+    public StrikethroughDelimiterProcessor() {
+        this(false);
+    }
+
+    public StrikethroughDelimiterProcessor(boolean requireTwoTildes) {
+        this.requireTwoTildes = requireTwoTildes;
+    }
+
     @Override
     public char getOpeningCharacter() {
         return '~';
@@ -22,7 +32,7 @@ public class StrikethroughDelimiterProcessor implements DelimiterProcessor {
 
     @Override
     public int getMinLength() {
-        return 1;
+        return requireTwoTildes ? 2 : 1;
     }
 
     @Override

--- a/commonmark-ext-gfm-strikethrough/src/test/java/org/commonmark/ext/gfm/strikethrough/StrikethroughTest.java
+++ b/commonmark-ext-gfm-strikethrough/src/test/java/org/commonmark/ext/gfm/strikethrough/StrikethroughTest.java
@@ -4,22 +4,25 @@ import org.commonmark.Extension;
 import org.commonmark.node.Node;
 import org.commonmark.node.Paragraph;
 import org.commonmark.node.SourceSpan;
+import org.commonmark.node.Text;
 import org.commonmark.parser.IncludeSourceSpans;
 import org.commonmark.parser.Parser;
+import org.commonmark.parser.delimiter.DelimiterProcessor;
+import org.commonmark.parser.delimiter.DelimiterRun;
 import org.commonmark.renderer.html.HtmlRenderer;
 import org.commonmark.renderer.text.TextContentRenderer;
 import org.commonmark.testutil.RenderingTestCase;
 import org.junit.Test;
 
 import java.util.Arrays;
-import java.util.Collections;
 import java.util.Set;
 
+import static java.util.Collections.singleton;
 import static org.junit.Assert.assertEquals;
 
 public class StrikethroughTest extends RenderingTestCase {
 
-    private static final Set<Extension> EXTENSIONS = Collections.singleton(StrikethroughExtension.create());
+    private static final Set<Extension> EXTENSIONS = singleton(StrikethroughExtension.create());
     private static final Parser PARSER = Parser.builder().extensions(EXTENSIONS).build();
     private static final HtmlRenderer HTML_RENDERER = HtmlRenderer.builder().extensions(EXTENSIONS).build();
     private static final TextContentRenderer CONTENT_RENDERER = TextContentRenderer.builder()
@@ -93,6 +96,19 @@ public class StrikethroughTest extends RenderingTestCase {
     }
 
     @Test
+    public void requireTwoTildesOption() {
+        Parser parser = Parser.builder()
+                .extensions(singleton(StrikethroughExtension.builder()
+                        .requireTwoTildes(true)
+                        .build()))
+                .customDelimiterProcessor(new SubscriptDelimiterProcessor())
+                .build();
+
+        Node document = parser.parse("~foo~ ~~bar~~");
+        assertEquals("(sub)foo(/sub) /bar/", CONTENT_RENDERER.render(document));
+    }
+
+    @Test
     public void sourceSpans() {
         Parser parser = Parser.builder()
                 .extensions(EXTENSIONS)
@@ -109,5 +125,30 @@ public class StrikethroughTest extends RenderingTestCase {
     @Override
     protected String render(String source) {
         return HTML_RENDERER.render(PARSER.parse(source));
+    }
+
+    private static class SubscriptDelimiterProcessor implements DelimiterProcessor {
+
+        @Override
+        public char getOpeningCharacter() {
+            return '~';
+        }
+
+        @Override
+        public char getClosingCharacter() {
+            return '~';
+        }
+
+        @Override
+        public int getMinLength() {
+            return 1;
+        }
+
+        @Override
+        public int process(DelimiterRun openingRun, DelimiterRun closingRun) {
+            openingRun.getOpener().insertAfter(new Text("(sub)"));
+            closingRun.getCloser().insertBefore(new Text("(/sub)"));
+            return 1;
+        }
     }
 }

--- a/commonmark/src/main/java/org/commonmark/internal/StaggeredDelimiterProcessor.java
+++ b/commonmark/src/main/java/org/commonmark/internal/StaggeredDelimiterProcessor.java
@@ -51,7 +51,7 @@ class StaggeredDelimiterProcessor implements DelimiterProcessor {
                 added = true;
                 break;
             } else if (len == pLen) {
-                throw new IllegalArgumentException("Cannot add two delimiter processors for char '" + delim + "' and minimum length " + len);
+                throw new IllegalArgumentException("Cannot add two delimiter processors for char '" + delim + "' and minimum length " + len + "; conflicting processors: " + p + ", " + dp);
             }
         }
         if (!added) {

--- a/commonmark/src/test/java/org/commonmark/test/DelimiterProcessorTest.java
+++ b/commonmark/src/test/java/org/commonmark/test/DelimiterProcessorTest.java
@@ -60,7 +60,7 @@ public class DelimiterProcessorTest extends RenderingTestCase {
     }
 
     @Test(expected = IllegalArgumentException.class)
-    public void multipleDelimitersWithSameLength() {
+    public void multipleDelimitersWithSameLengthConflict() {
         Parser.builder()
                 .customDelimiterProcessor(new OneDelimiterProcessor())
                 .customDelimiterProcessor(new OneDelimiterProcessor())

